py-rse/py-scripting.Rmd

```diff
@@ -135,15 +129,399 @@ optional arguments:
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument('infile', type=argparse.FileType('r'), nargs='?',
+                        default='-', help='Input file name')
+    parser.add_argument('-n', type=int, default=None,
+                        help='Limit output to n most frequent words')
+    args = parser.parse_args()
+    main(args)
+```
+
+and here's `collate.py`:
+
+```python
+"""Combine multiple word count CSV-files into a single cumulative count."""
+
+import sys
+import csv
+import argparse
+from collections import Counter
+import mymodule
+
+def update_counts(reader, word_counts):
+    """Update word counts with data from another reader/file."""
+    for word, count in csv.reader(reader):
+        word_counts[word] += int(count)
+
+def main(args):
+    """Run the command line program."""
+    word_counts = Counter()
+    for fn in args.infiles:
+        with open(fn, 'r') as reader:
+            update_counts(reader, word_counts)
+    word_counts = word_counts.most_common()
+    mymodule.report_results(word_counts, top_n=args.n)
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument('infiles', type=str, nargs='*', help='Input file names')
+    parser.add_argument('-n', type=int, default=None, help='Limit output to n most frequent words')
+    args = parser.parse_args()
+    main(args)
+```
+
+## Mission Critical Exercise {#py-rse-py-scripting-critical-exercise}
+
+The last thing for us to do is to plot the word count distribution.
+
+Recall that [Zipf's law][zipfs-law] states the second most common word in a body of text
+appears half as often as the most common,
+the third most common appears a third as often, and so on.
+Mathematically, this might be written as
+"word frequency is proportional to 1/rank."
+
+The following code plots the word frequency against the inverse rank 
+using the popular pandas library:
+
+```python
+import pandas
+import matplotlib.pyplot as plt
+
+input_csv = 'emma_words.csv'
+df = pd.read_csv(input_csv, header=None, names=('word', 'word_frequency'))
+df['rank'] = df['word_frequency'].rank(ascending=False)
+df['inverse_rank'] = 1 / df['rank'] 
+df.plot.scatter(x='word_frequency', y='inverse_rank',
+                figsize=[12, 6], grid=True)
+plt.show()
+```
+
+```{r py-rse-py-scripting-repl, echo=FALSE, fig.cap="Word frequency distribution for the book Emma"}
+knitr::include_graphics("figures/py-rse-scripting/emma_words.png")
+```
+
+Using `script_template.py` as a guide,
+take this plotting code and write a new python script called `plotcounts.py`.
+The script should:
+
+a. Use the `type=argparse.FileType('r')`, `nargs='?'` and `default='-'` options
+   for the input file argument (i.e. similar to the `countwords.py` script)
+   so that `plotcounts.py` uses standard input if no csv file is given.
+
+b. Add an optional `--xlim` option so that the user can change the x-axis bounds.
```

Comment:
It doesn't seem to have come up in this chapter but why two dashes here?  

The "Key points" talk about using short and long command line options, but that doesn't seem to be explicitly talked about in the chapter.  Might also need to comment on when to use positional vs. optional arguments.  Does `argparse` do anything different based on the name you give an argument, i.e. if it has no, one or two dashes?

Reply:
Another thought that occurred as I started on command line git: do you ever talk about where one might be saving these python programs? Do their relative locations matter?

Reply:
@SaraMati had the same question in relation to the shell lessons. I think we're ultimately going to have to have a zip file that people download and unzip on their Desktop with all the data files and various directories we use in it. In the meantime, I think we can just assume they are saving the scripts in the `/zipfs_law/bin/` directory.

Reply:
I've removed the reference to long and short options from the key points and I've created a call out box to talk about the different between positional and optional arguments (and short and long optional arguments).